pkg/controller/shared_state.go

```diff
@@ -1565,3 +1573,62 @@ func (c *SharedState) getNodesTopologySelector(ctx context.Context, nodes []v1.N
 	}
 	return topologySelector
 }
+
+func (c *SharedState) addNodeToAvailableAttachmentsMap(ctx context.Context, nodeName string, nodeLables map[string]string) {
+	if _, ok := c.availableAttachmentsMap.Load(nodeName); !ok {
+		capacity, err := azureutils.GetNodeRemainingDiskCount(ctx, c.cachedClient, nodeName)
+		if err != nil {
+			klog.Errorf("Failed to get node(%s) remaining disk count with error: %v", nodeName, err)
+			// store the maximum capacity if an entry for the node doesn't exist.
+			capacity, err = azureutils.GetNodeMaxDiskCountWithLabels(nodeLables)
+			if err != nil {
+				klog.Errorf("Failed to add node(%s) in availableAttachmentsMap, because get capacity of available attachments is failed with error: %v", nodeName, err)
+				return
+			}
+		}
+		var count atomic.Int32
+		count.Store(int32(capacity))
+		klog.Infof("Added node(%s) to availableAttachmentsMap with capacity: %d", nodeName, capacity)
+		c.availableAttachmentsMap.LoadOrStore(nodeName, &count)
+	}
+
+	klog.Infof("Node(%s) is already in availableAttachmentsMap. Don't need to add it again", nodeName)
+}
+
+func (c *SharedState) deleteNodeFromAvailableAttachmentsMap(ctx context.Context, node string) {
+	klog.Infof("Deleted node(%s) from availableAttachmentsMap", node)
+	c.availableAttachmentsMap.Delete(node)
+}
+func (c *SharedState) decrementAttachmentCount(ctx context.Context, node string) bool {
+	remainingCapacity, nodeExists := c.availableAttachmentsMap.Load(node)
+	if nodeExists && remainingCapacity != nil {
+		currentCapacity := int32(0)
+		for {
+			currentCapacity = remainingCapacity.(*atomic.Int32).Load()
+			if currentCapacity == int32(0) || remainingCapacity.(*atomic.Int32).CompareAndSwap(currentCapacity, currentCapacity-1) {
+				if currentCapacity == int32(0) {
+					klog.Errorf("Failed to decrement attachment count for node(%s), because no available attachment", node)
+					return false
+				}
+				klog.Infof("Decrement attachment count is succeeded for node(%s) with new capacity: %d", node, currentCapacity-1)
```

Comment:
This log isn't all that useful and can be removed.

Reply:
Done.